bin/src/upgrade.rs

```diff
@@ -54,17 +55,20 @@ pub struct UpgradeData {
   //pub token_count: usize,
 }
 
-pub fn start_new_main_process(executable_path: String, upgrade_data: UpgradeData) -> (pid_t, Channel<(),bool>) {
+pub fn start_new_main_process(
+  executable_path: String,
+  upgrade_data: UpgradeData
+) -> Result<(pid_t, Channel<(),bool>), anyhow::Error> {
   trace!("parent({})", unsafe { libc::getpid() });
 
-  let mut upgrade_file = tempfile().expect("could not create temporary file for upgrade");
+  let mut upgrade_file = tempfile().context("could not create temporary file for upgrade")?;
```

Comment:
if you could switch from the `context` method to the `with_context` method, it will help to not allocate each time the error, even though it is a success